R/GeospatialFunctions.R

```diff
@@ -0,0 +1,571 @@
+      urltools::param_set(key = "f", value = "geojson") %>%
+      urltools::param_set(key = "outFields", value = "*") %>%
+      urltools::param_set(key = "geometryType", value = "esriGeometryEnvelope") %>%
+      urltools::param_set(key = "returnGeometry", value = "true") %>%
+      urltools::param_set(key = "returnTrueCurves", value = "false") %>%
+      urltools::param_set(key = "returnIdsOnly", value = "false") %>%
+      urltools::param_set(key = "returnCountOnly", value = "false") %>%
+      urltools::param_set(key = "returnZ", value = "false") %>%
+      urltools::param_set(key = "returnM", value = "false") %>%
+      urltools::param_set(key = "returnDistinctValues", value = "false") %>%
+      urltools::param_set(key = "returnExtentOnly", value = "false") %>%
+      urltools::param_set(key = "featureEncoding", value = "esriDefault")
+    
+    # Fetch features within the offset window and append to list:
+    features <- suppressMessages(suppressWarnings({tryCatch({
+      geojsonsf::geojson_sf(query)
+    }, error = function(e) {
+      NULL
+    })}))
+    
+    
+    # Exit loop if no more features or error occurred
+    if (is.null(features) || nrow(features) == 0) {
+      break
+    }
+    
+    all_features <- c(all_features, list(features))
+    # once done, change offset by 2000 features:
+    offset <- offset + 2000
+    
+    if(offset == 4000){print("Your TADA data covers a large spatial range. The ATTAINS pull may take a while.")}
+    
+  }
+  
+  all_features <- bind_rows(all_features)
+  
+  return(all_features)
+  
+}
+
+
+#' TADA_GetATTAINS
+#' 
+#' Link catchment-based ATTAINS assessment unit data to Water Quality Portal observations, often imported via `TADA_DataRetrieval()`. This function returns the same raw objects that are mapped in `TADA_ViewATTAINS()`.
+#' 
+#' @param data A dataframe created by `TADA_DataRetrieval()` or the {sf} equivalent made by `TADA_MakeSpatial()`.
+#' @param return Whether to add the associated ATTAINS_catchments, ATTAINS_lines, ATTAINS_points, and ATTAINS_polygons shapefile objects into your Global Environment. TRUE (yes, return) or FALSE (no, do not return). All ATTAINS features are in WGS84 (crs = 4326).
+#' 
+#' @return The original `TADA_DataRetrieval()` dataframe with additional columns associated with the ATTAINS assessment unit data. 
+#' 
+#' @seealso [TADA_DataRetrieval()]
+#' @seealso [TADA_MakeSpatial()]
+#' @seealso [TADA_ViewATTAINS()]
+#' 
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#'tada_data <- TADA_DataRetrieval(startDate = "1990-01-01",
+#'                                endDate = "1995-12-31",
+#'                                characteristicName = "pH",
+#'                                statecode = "NV",
+#'                                applyautoclean = TRUE)
+#'                                  
+#'tada_attains <- TADA_GetATTAINS(data = tada_data, return = TRUE)
+#' }
+
+TADA_GetATTAINS <- function(data, return = FALSE){
```

Comment:
This function also requires certain columns to be present but there are no checks for those columns like there is in `TADA_MakeSpatial()`